src/outputs/parthenon_opmd.cpp

```diff
@@ -0,0 +1,517 @@
+  return {mesh_record_name, comp_name};
+}
+
+std::tuple<openPMD::Offset, openPMD::Extent>
+GetChunkOffsetAndExtent(Mesh *pm, std::shared_ptr<MeshBlock> pmb) {
+  openPMD::Offset chunk_offset;
+  openPMD::Extent chunk_extent;
+  const auto loc = pm->Forest().GetLegacyTreeLocation(pmb->loc);
+  if (pm->ndim == 3) {
+    chunk_offset = {loc.lx3() * static_cast<uint64_t>(pmb->block_size.nx(X3DIR)),
+                    loc.lx2() * static_cast<uint64_t>(pmb->block_size.nx(X2DIR)),
+                    loc.lx1() * static_cast<uint64_t>(pmb->block_size.nx(X1DIR))};
+    chunk_extent = {static_cast<uint64_t>(pmb->block_size.nx(X3DIR)),
+                    static_cast<uint64_t>(pmb->block_size.nx(X2DIR)),
+                    static_cast<uint64_t>(pmb->block_size.nx(X1DIR))};
+  } else if (pm->ndim == 2) {
+    chunk_offset = {loc.lx2() * static_cast<uint64_t>(pmb->block_size.nx(X2DIR)),
+                    loc.lx1() * static_cast<uint64_t>(pmb->block_size.nx(X1DIR))};
+    chunk_extent = {static_cast<uint64_t>(pmb->block_size.nx(X2DIR)),
+                    static_cast<uint64_t>(pmb->block_size.nx(X1DIR))};
+  } else {
+    PARTHENON_THROW("1D output for openpmd not yet supported.");
+  }
+  return {chunk_offset, chunk_extent};
+}
+} // namespace OpenPMDUtils
+
+//----------------------------------------------------------------------------------------
+//! \fn void OpenPMDOutput:::WriteOutputFile(Mesh *pm)
+//  \brief  Expose mesh and all Cell variables for processing with Ascent
+void OpenPMDOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm,
+                                    const SignalHandler::OutputSignal signal) {
+#ifndef PARTHENON_ENABLE_OPENPMD
+  if (Globals::my_rank == 0) {
+    PARTHENON_WARN("OpenPMD output requested by input file, but OpenPMD support not "
+                   "compiled in. Skipping this output type.");
+  }
+#else
+  using openPMD::Access;
+  using openPMD::Series;
+
+  // TODO(pgrete) .h5 for hd5 and .bp for ADIOS2 or .json for JSON
+  // TODO(pgrete) check if CREATE is the correct pattern (for not overwriting the series
+  // but an interation) This just describes the pattern of the filename. The correct file
+  // will be accessed through the iteration idx below. The file suffix maps to the chosen
+  // backend.
+  // TODO(pgrete) add final and now logic
+  Series series =
+      Series(output_params.file_basename + "." + output_params.file_id + ".%05T.bp",
+             Access::CREATE);
+
+  // TODO(pgrete) How to handle downstream info, e.g.,  on how/what defines a vector?
+  // TODO(pgrete) Should we update for restart or only set this once? Or make it per
+  // iteration?
+  // ... = pin->GetString(output_params.block_name, "actions_file");
+  series.setAuthor("My Name <mail@addre.es");
+  series.setComment("Hello world!");
+  series.setMachine("bla");
+  series.setSoftware("Parthenon + Downstream info");
+  series.setDate("2024-02-29");
+
+  // TODO(pgrete) Units?
```

Comment:
Interesting... How have different teams/codes been handling units? In Phoebus we have a fairly robust units infrastructure that @brryan put together. We could potentially upstream it.

Reply:
Does Phoebus use opaque types or some other magic to ensure dimensional compatibility at compile-time?
(The AthenaPK units infrastructure is here: https://github.com/parthenon-hpc-lab/athenapk/blob/2bee272ff09f43dc942188bc522c061ba50a1643/src/units.hpp#L16.)

For full compliance, the OpenPMD standard requires a conversion factor for each field to go from data units -> SI units: https://github.com/openPMD/openPMD-standard/blob/latest/STANDARD.md#unit-systems-and-dimensionality. For dimensionless or relativistic units, this is somewhat annoying. The recommendation in the standard is to provide a conversion factor for a reference dimensional scaling.